text/3855-mitigation-enforcement.md

```diff
@@ -0,0 +1,448 @@
+3. [`-Z no-jump-tables`] - CFI-type mitigation?
+   soon to be stabilized as [`-C jump-tables`]. Do we want to hold that
+   stabilization as well?
+3. [`-Z retpoline`] and `-Z retpoline-external-thunk` - Spectre-type mitigation
+4. [`-Z sanitizer`]-based mitigations. A fairly large use case.
+   As far as I can tell, enforcement makes sense for
+   `-Zsanitizer=cfi, -Zsanitizer=memtag, -Zsanitizer=shadow-call-stack`
+   and does not make sense for
+   `-Zsanitizer=address, -Zsanitizer=dataflow, -Zsanitizer=hwaddress, -Zsanitizer=leak, -Zsanitizer=memory, -Zsanitizer=thread`
+4. [`-Z stack-protector`] - stack smashing protection, local-type mitigation
+5. [`-Z ub-checks`]: as far as I can tell, it is not intended as a mitigation,
+   but since it prevents some UB, it might be thought of as one
+
+[`-Z branch-protection`]: https://github.com/rust-lang/rust/issues/113369
+[`-Z cf-protection`]: https://github.com/rust-lang/rust/issues/93754
+[`-Z indirect-branch-cs-prefix`]: https://github.com/rust-lang/rust/issues/116852
+[`-Z no-jump-tables`]: https://github.com/rust-lang/rust/issues/116592
+[`-C jump-tables`]: https://github.com/rust-lang/rust/pull/145974
+[`-Z retpoline`]: https://github.com/rust-lang/rust/issues/116852
+[`-Z sanitizer`]: https://github.com/rust-lang/rust/issues/89653
+[`-Z stack-protector`]: https://github.com/rust-lang/rust/issues/114903
+[`-Z ub-checks`]: https://github.com/rust-lang/rust/issues/123499
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When you use a mitigation, such as `-C stack-protector=strong`, if one of your
+dependencies does not have that mitigation enabled, compilation will fail.
+
+> Error: your program uses the crate `foo`, that is not protected by
+> `-C stack-protector=strong`.
+>
+> Recompile that crate with the mitigation enabled, or use
+> `-C stack-protector=strong-noenforce` to allow creating an artifact
+> that has the mitigation only partially enabled.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Every flag value that enables a mitigation for which enforcement is desired is split
+into 2 separate values, "enforcing" and "non-enforcing" mode. The enforcing mode
+is the default, non-enforcing mode is constructed by adding `-noenforce` to the
+name of the value, for example `-C stack-protector=strong-noenforce` or
+`-C sanitizer=shadow-call-stack-noenforce`.
+
+> It is possible to bikeshed the exact naming scheme.
+
+> Every new mitigation would need to decide whether it adopts this scheme,
+> but mitigations are expected to adopt it.
+
+Every crate gets a metadata field that contains the set of mitigations it has enabled.
+
+When compiling a crate, if the current crate has a mitigation with enforcement
+turned on, and one of the dependencies does not have that mitigation turned
+on (whether enforcing or not), a compilation error results.
+
+If a mitigation has multiple "levels", a lower level at a dependent
```

Comment:
This might need to be clarified. Do you mean a stricter or less strict level in the dependent crate?

It might also be helpful to use the same dependency terms as the previous paragraph (current and dependency).